source/lac/solver_gmres.cc

```diff
@@ -53,61 +53,74 @@ namespace internal
           constexpr unsigned int inner_batch_size =
             delayed_reorthogonalization ? 6 : 12;
 
-          for (; c < locally_owned_size / n_lanes / inner_batch_size;
-               ++c, j += n_lanes * inner_batch_size)
-            {
-              VectorizedArray<double> vvec[inner_batch_size];
-              for (unsigned int k = 0; k < inner_batch_size; ++k)
-                vvec[k].load(current_vector + j + k * n_lanes);
-              VectorizedArray<double> prev_vector[inner_batch_size];
-              for (unsigned int k = 0; k < inner_batch_size; ++k)
-                prev_vector[k].load(orthogonal_vectors[n_vectors - 1] + j +
-                                    k * n_lanes);
-
-              {
-                VectorizedArray<double> local_sum_0 = prev_vector[0] * vvec[0];
-                VectorizedArray<double> local_sum_1 =
-                  prev_vector[0] * prev_vector[0];
-                VectorizedArray<double> local_sum_2 = vvec[0] * vvec[0];
-                for (unsigned int k = 1; k < inner_batch_size; ++k)
-                  {
-                    local_sum_0 += prev_vector[k] * vvec[k];
-                    if (delayed_reorthogonalization)
-                      {
-                        local_sum_1 += prev_vector[k] * prev_vector[k];
-                        local_sum_2 += vvec[k] * vvec[k];
-                      }
-                  }
-                hs[n_vectors - 1] += local_sum_0;
-                if (delayed_reorthogonalization)
-                  {
-                    correct[n_vectors - 1] += local_sum_1;
-                    correct[n_vectors] += local_sum_2;
-                  }
-              }
-
-              for (unsigned int i = 0; i < n_vectors - 1; ++i)
+          const unsigned int loop_length_c =
+            locally_owned_size / n_lanes / inner_batch_size;
+          for (unsigned int c_block = 0; c_block < (loop_length_c + 63) / 64;
+               ++c_block)
+            for (unsigned int i_block = 0; i_block < (n_vectors + 7) / 8;
+                 ++i_block)
+              for (c = c_block * 64, j = c * n_lanes * inner_batch_size;
+                   c < std::min(loop_length_c, (c_block + 1) * 64);
+                   ++c, j += n_lanes * inner_batch_size)
```

Comment:
Would you mind prefixing all of this with a description of why you break it into three loops? And perhaps include the original one loop into the comments?

Reply:
Excellent idea, I will do soon!

Reply:
I added a long comment in one function, and a shorter one with reference to the first comment in the second place I adjusted. Thanks for the discussion, I think this helps quite a bit.